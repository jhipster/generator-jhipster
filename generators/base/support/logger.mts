/**
 * Copyright 2013-2023 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import chalk from 'chalk';
import type TerminalAdapter from 'yeoman-environment/lib/adapter';
import createDebug, { type Debugger } from 'debug';

/**
 * formats the message to be displayed in the console.
 * @param message the info message to format.
 */
const formatWarningMessageHeader = message => {
  return `${chalk.yellow.bold('WARNING!')} ${message}`;
};

const formatErrorMessageHeader = message => {
  return `${chalk.red.bold('ERROR!')} ${message}`;
};

const formatFatalMessageHeader = message => {
  return `${chalk.red.bold('FATAL!')} ${message}`;
};

const formatInfoMessageHeader = message => {
  return `${chalk.green('INFO!')} ${message}`;
};

export type LoggerOptions = {
  adapter: TerminalAdapter;
  namespace?: string;
  debugEnabled?: boolean;
};

export const CLI_LOGGER = 'jhipster:cli';

export default class Logger {
  adapter: TerminalAdapter;
  debugger: Debugger;
  debugEnabled: boolean;

  constructor({ adapter, namespace = 'jhipster', debugEnabled }: LoggerOptions) {
    this.adapter = adapter;
    this.debugger = createDebug(namespace);

    const cliLogger = namespace === CLI_LOGGER;
    if (cliLogger) {
      this.debugEnabled = debugEnabled || process.argv.includes('-d') || process.argv.includes('--debug'); // Need this early
    } else {
      this.debugEnabled = debugEnabled ?? false;
    }
    if (debugEnabled) {
      this.debugger.enabled = true;
    }
    if (cliLogger && this.debugEnabled) {
      this.info('Debug logging is on');
    }
  }

  debug(msg, ...args) {
    this.debugger(msg, ...args);
  }

  warn(msg) {
    const warn = formatWarningMessageHeader(msg);
    this.adapter.log(warn);
  }

  info(msg) {
    const info = formatInfoMessageHeader(msg);
    this.adapter.log(info);
  }

  log(msg) {
    this.adapter.log(msg);
  }

  error(msg) {
    const errorMessage = formatErrorMessageHeader(msg);
    this.adapter.log(errorMessage);

    process.exitCode = 1;
  }

  /**
   *  Use with caution.
   *  process.exit is not recommended by Node.js.
   *  Refer to https://nodejs.org/api/process.html#process_process_exit_code.
   */
  fatal(msg, trace) {
    const fatalMessage = formatFatalMessageHeader(msg);
    this.adapter.log(fatalMessage);
    if (trace) {
      this.adapter.log(trace);
    }

    process.exit(1);
  }
}
