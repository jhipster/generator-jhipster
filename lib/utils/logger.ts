/**
 * Copyright 2013-2025 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import chalk from 'chalk';
import type { Logger as LoggerApi } from '@yeoman/types';
import type { Logger as DefaultLogger, LoggerOptions } from '@yeoman/adapter';
import { createLogger } from '@yeoman/adapter';
import createDebug from 'debug';

/**
 * formats the message to be displayed in the console.
 * @param message the info message to format.
 */
const addPrefixToParameters = (prefix: string, ...args: Parameters<LoggerApi['write']>): Parameters<LoggerApi['write']> => {
  args[0] = `${prefix} ${args[0]}`;
  return args;
};

const formatWarningMessageHeader = (...args: Parameters<LoggerApi['write']>): Parameters<LoggerApi['write']> =>
  addPrefixToParameters(chalk.yellow.bold('WARNING!'), ...args);

const formatErrorMessageHeader = (...args: Parameters<LoggerApi['write']>): Parameters<LoggerApi['write']> =>
  addPrefixToParameters(chalk.red.bold('ERROR!'), ...args);

const formatFatalMessageHeader = (...args: Parameters<LoggerApi['write']>): Parameters<LoggerApi['write']> =>
  addPrefixToParameters(chalk.red.bold('FATAL!'), ...args);

const formatInfoMessageHeader = (...args: Parameters<LoggerApi['write']>): Parameters<LoggerApi['write']> =>
  addPrefixToParameters(chalk.green('INFO!'), ...args);

export const CLI_LOGGER = 'jhipster:cli';

export const createJHipsterLogger = (options: LoggerOptions & { namespace?: string; debugEnabled?: boolean } = {}) => {
  const { namespace = 'jhipster' } = options;
  const debug = createDebug(namespace);

  const customJHipsterLogger = {
    debugger: debug,

    debug(msg, ...args) {
      this.debugger(msg, ...args);
    },

    warn(this: LoggerApi, ...args: Parameters<LoggerApi['write']>) {
      this.writeln(...formatWarningMessageHeader(...args));
      return this;
    },

    verboseInfo(this: LoggerApi, ...args: Parameters<LoggerApi['write']>) {
      this.writeln(...formatInfoMessageHeader(...args));
      return this;
    },

    log(this: LoggerApi, ...args: Parameters<LoggerApi['write']>) {
      this.writeln(...args);
      return this;
    },

    error(this: DefaultLogger, msg, error) {
      const errorMessage = formatErrorMessageHeader(msg);
      this.console.error(...errorMessage);
      if (error) {
        this.console.error(error);
      }

      process.exitCode = 1;
    },

    fatal(this: any, msg, trace?) {
      const fatalMessage = formatFatalMessageHeader(msg);
      this.console.error(...fatalMessage);
      if (trace) {
        this.console.error(trace);
      }

      process.exit(1);
    },
  };

  const logger = createLogger({ ...options, loggers: customJHipsterLogger });
  const cliLogger = namespace === CLI_LOGGER;
  let debugEnabled = options.debugEnabled;
  if (cliLogger) {
    debugEnabled = debugEnabled || process.argv.includes('-d') || process.argv.includes('--debug'); // Need this early
    if (debugEnabled) {
      logger.verboseInfo('Debug logging is on');
    }
  }
  if (debugEnabled) {
    logger.debugger.enabled = true;
  }
  return logger;
};

export type Logger = ReturnType<typeof createJHipsterLogger>;
