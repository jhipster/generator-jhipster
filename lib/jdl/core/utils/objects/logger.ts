/**
 * Copyright 2013-2025 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { createJHipsterLogger } from '../../../../utils/logger.js';

const logger = createJHipsterLogger({ namespace: 'jhipster:jdl' });

export default {
  info: (message: string) => logger.verboseInfo(message),
  error: (message: string) => logger.error(message),
  warn: (message: string) => logger.warn(message),
  debug: (message: string) => logger.debug(message),
};
