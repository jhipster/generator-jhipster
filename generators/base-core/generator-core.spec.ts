/* eslint-disable no-unused-expressions */
import { it, describe, expect as jestExpect, beforeEach } from 'esmocha';
import { basicHelpers as helpers } from '../../test/support/index.js';

import Base from './index.js';
import { createJHipsterLogger } from '../base/support/logger.js';

const BaseGenerator: any = Base.prototype;

BaseGenerator.log = msg => {
  // eslint-disable-next-line no-console
  console.log(msg);
};

BaseGenerator.logger = createJHipsterLogger();

describe('generator - base-core', () => {
  describe('passing arguments', () => {
    let Dummy;
    beforeEach(async () => {
      await helpers.prepareTemporaryDir();
      Dummy = helpers.createDummyGenerator(Base);
    });

    it('no argument', async () => {
      const base = new Dummy([], { sharedData: {}, env: await helpers.createTestEnv() });
      base.parseJHipsterArguments({
        jdlFiles: {
          type: String,
        },
      });
      jestExpect(base.jdlFiles).toBe(undefined);
    });
    it('undefined positional arguments', async () => {
      const base = new Dummy({ positionalArguments: [], sharedData: {}, env: await helpers.createTestEnv() });
      base.parseJHipsterArguments({
        jdlFiles: {
          type: String,
        },
      });
      jestExpect(base.jdlFiles).toBe(undefined);
    });
    it('undefined argument', async () => {
      const base = new Dummy([undefined], { sharedData: {}, env: await helpers.createTestEnv() });
      base.parseJHipsterArguments({
        jdlFiles: {
          type: String,
        },
      });
      jestExpect(base.jdlFiles).toBe(undefined);
    });
    it('undefined positional arguments', async () => {
      const base = new Dummy({ positionalArguments: [undefined], sharedData: {}, env: await helpers.createTestEnv() });
      base.parseJHipsterArguments({
        jdlFiles: {
          type: String,
        },
      });
      jestExpect(base.jdlFiles).toBe(undefined);
    });
    it('string arguments', async () => {
      const base = new Dummy(['foo'], { sharedData: {}, env: await helpers.createTestEnv() });
      base.parseJHipsterArguments({
        jdlFiles: {
          type: String,
        },
      });
      jestExpect(base.jdlFiles).toBe('foo');
    });
    it('vararg arguments', async () => {
      const base = new Dummy(['bar', 'foo'], { sharedData: {}, env: await helpers.createTestEnv() });
      base.parseJHipsterArguments({
        first: {
          type: String,
        },
        jdlFiles: {
          type: Array,
        },
      });
      jestExpect(base.first).toBe('bar');
      jestExpect(base.jdlFiles).toMatchObject(['foo']);
    });
    it('vararg arguments using positionalArguments', async () => {
      const base = new Dummy({ positionalArguments: ['bar', ['foo']], sharedData: {}, env: await helpers.createTestEnv() });
      base.parseJHipsterArguments({
        first: {
          type: String,
        },
        jdlFiles: {
          type: Array,
        },
      });
      jestExpect(base.first).toBe('bar');
      jestExpect(base.jdlFiles).toMatchObject(['foo']);
    });
  });
});
