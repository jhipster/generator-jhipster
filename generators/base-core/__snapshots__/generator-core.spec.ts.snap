// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`generator - base-core editPropertiesFile passing sortFile option as false when properties file exists should append to an existing property 1`] = `
{
  "dummy.properties": {
    "contents": "append.existing = existingValue, appendedValueUsingFunction
# comment will be kept for no sortFile
override.existing = overriddenValue
new.property = newValue
# comment added
append.separator = firstSeparator, secondSeparator",
    "stateCleared": "modified",
  },
}
`;

exports[`generator - base-core editPropertiesFile passing sortFile option as true when properties file exists should append to an existing property 1`] = `
{
  "dummy.properties": {
    "contents": "append.existing = existingValue, appendedValueUsingFunction
append.separator = firstSeparator, secondSeparator
new.property = newValue
override.existing = overriddenValue",
    "stateCleared": "modified",
  },
}
`;
